### PR TITLE
C Strings

### PR DESCRIPTION
**Pull request by jjgarzella**
These commits add support for C Strings, along with regression tests for each string method (where it makes sense).

### CODE REVIEW

**Comment by michael-emmi** [comment]
Perhaps we should separate the definition of functions declared in `string.h` to a separate compilation unit, e.g., called `string.c`. We can put this in `share/smack/lib/string.c` right next to `share/smack/lib/smack.c` and `share/smack/lib/pthread.c` and only compile/link it when a certain flag is passed to Smack.

I also think that at this point we ought to think of standardizing the command line flags for including stubs. Currently we are providing stubs for `string.h` and `pthread.h` via flags `--strings` and `--pthread`. Why don’t we just provide a generic flag `--stub <library-name>` which would be used as `--stub string` and `--stub pthread`.

**Comment by michael-emmi** [comment]
Has this been dropped?

This is a useful addition that we should merge.

 But it should certainly be rebased. Also string function definitions should be separated from Smack and `__VERIFIER` function definitions, i.e., into `share/smack/lib/string.c`. Also the test directory should probably be called `string` or `strings` and not `cstrings`. And there should not be a need to use a `STRINGS_ENABLED` compiler directive, since the decision to include the compiled `share/smack/lib/string.c` module would be taken by the top-level Smack script.

**Comment by jjgarzella** [comment]
No, this has not been dropped. I think right now we're waiting for Zvonimir to review it and give his opinion. I have also been looking around for benchmarks to test this with, and may end up writing some of my own.

File: share/smack/lib/smack.c [hunk 1]
```diff
@@ -2228,19 +2228,95 @@ void __SMACK_init_func_memory_model(void) {
 }
 
 
-#if MEMORY_SAFETY || SIGNED_INTEGER_OVERFLOW_CHECK
+#if STRINGS_ENABLED || (MEMORY_SAFETY || SIGNED_INTEGER_OVERFLOW_CHECK)
```
Comment (michael-emmi) [inline share/smack/lib/smack.c]: This compiler directive should not exist. String functions should be extracted to a separate file `share/smack/lib/string.c` which is conditionally included by the Smack scripts.

File: share/smack/lib/smack.c [hunk 2]
```diff
@@ -2228,19 +2228,95 @@ void __SMACK_init_func_memory_model(void) {
...
 char *strcpy(char *dest, const char *src) {
   char *save = dest;
   while (*dest++ = *src++);
   return save;
 }
 
+char *strncpy(char *dest, const char *src, size_t n) {
```
Comment (michael-emmi) [inline share/smack/lib/smack.c]: String functions should be extracted to a separate file `share/smack/lib/string.c` which is conditionally included by the Smack scripts.

File: share/smack/top.py [hunk 3]
```diff
@@ -318,6 +320,7 @@ def default_clang_compile_command(args, lib = False):
   if args.memory_safety: cmd += ['-DMEMORY_SAFETY']
   if args.signed_integer_overflow: cmd += (['-ftrapv'] if not lib else ['-DSIGNED_INTEGER_OVERFLOW_CHECK'])
   if args.float: cmd += ['-DFLOAT_ENABLED']
+  if args.strings: cmd += ['-DSTRINGS_ENABLED']
```
Comment (michael-emmi) [inline share/smack/top.py]: This compiler directive should not exist. String functions should be extracted to a separate file `share/smack/lib/string.c` which is conditionally included by the Smack scripts.

File: test/cstrings/config.yml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+skip: false
```
Comment (michael-emmi) [inline test/cstrings/config.yml]: Not sure why this directory should be called `cstrings` rather than `string` or `strings`.

**Comment by michael-emmi** [comment]
Ok @jjgarzella we should merge this soon. Let’s first wait until #265 is resolved and merge to be consistent with the decisions made there, and avoid conflicts. Hopefully this can be finished this week :-)

File: test/strings/strcat.c [hunk 5]
```diff
@@ -0,0 +1,19 @@
+
+#include <string.h>
+#include "smack.h"
+
+// @flag --unroll=5
```
Comment (zvonimir) [inline test/strings/strcat.c]: This is the default unroll for this folder as you specified in `config.yml`. So remove this line from regressions that require unroll of 5. It is redundant.

File: test/strings/strcat.c [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+// @expect verified
+
+int main() {
```
Comment (zvonimir) [inline test/strings/strcat.c]: This should be `int main(void)`. Please fix this everywhere.

File: test/strings/strcat.c [hunk 7]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (zvonimir) [inline test/strings/strcat.c]: No need for this new line. You should simply start with `#include`. Fix this everywhere.

File: test/strings/strtok.c [hunk 8]
```diff
@@ -0,0 +1,21 @@
+
+#include <string.h>
+#include "smack.h"
+
+// @flag --unroll=5
+// @expect verified
+
+int main() {
+
```
Comment (zvonimir) [inline test/strings/strtok.c]: Remove this new line.

File: test/strings/strtok_fail.c [hunk 9]
```diff
@@ -0,0 +1,21 @@
+
+#include <string.h>
+#include "smack.h"
+
+// @flag --unroll=5
+// @expect error
+
+int main() {
+
```
Comment (zvonimir) [inline test/strings/strtok_fail.c]: Remove this new line.

File: share/smack/lib/string.c [hunk 10]
```diff
@@ -10,12 +10,88 @@ char *strcpy(char *dest, const char *src) {
...
+    s2++;
+  }
+  
+  return 0;
+}
+
+// concatenation
+
+char *strcat(char *dest, const char *src) {
+
```
Comment (zvonimir) [inline share/smack/lib/string.c]: Remove this new line.

File: share/smack/lib/string.c [hunk 11]
```diff
@@ -10,12 +10,88 @@ char *strcpy(char *dest, const char *src) {
...
+  size_t n;
+  for (n = 0; s1[n] == s2[n]; n++)
+    if (s1[n] == '\0')
+      return 0;
+  return s1[n] - s2[n];
+}
+
+int strncmp(const char *s1, const char *s2, size_t n) {
+ 
+// what do we do if n is bigger than s1 and s2?
```
Comment (zvonimir) [inline share/smack/lib/string.c]: It is fine to remove this comment. This is a known problem in real-world implementations.

File: share/smack/lib/string.c [hunk 12]
```diff
@@ -10,12 +10,88 @@ char *strcpy(char *dest, const char *src) {
...
+  while (*dest)
+    dest++;
+  while (*dest++ = *src++) ;
+
+  return retDest;
+}
+
+char *strncat(char *dest, const char *src, size_t n) {
+  
+ // what happens when n is too big?
```
Comment (zvonimir) [inline share/smack/lib/string.c]: It is fine to remove this comment. This is a known problem in real-world implementations.

File: share/smack/lib/string.c [hunk 13]
```diff
@@ -10,12 +10,88 @@ char *strcpy(char *dest, const char *src) {
...
+
```
Comment (zvonimir) [inline share/smack/lib/string.c]: Remove this new line.

File: share/smack/top.py [hunk 14]
```diff
@@ -304,6 +305,7 @@ def default_clang_compile_command(args, lib = False):
...
```
Comment (zvonimir) [inline share/smack/top.py]: This is not the right way to turn on strings. Take a look at: https://github.com/smackers/smack/blob/develop/share/smack/top.py#L317

File: share/smack/top.py [hunk 15]
```diff
@@ -181,6 +181,7 @@ def arguments():
   translate_group.add_argument('--split-aggregate-values', action='store_true', default=False,
     help='enable splitting of load/store instructions of LLVM aggregate types')
 
+  translate_group.add_argument('--strings', action='store_true', default=False, help='enable c string methods')
```
Comment (zvonimir) [inline share/smack/top.py]: say `enable support for strings`

File: test/strings/strcat.c [hunk 16]
```diff
@@ -0,0 +1,17 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char brick[10];
+  strcpy(brick,"bl");
+
```
Comment (zvonimir) [inline test/strings/strcat.c]: No need for this new line.

File: test/strings/strcat.c [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char brick[10];
+  strcpy(brick,"bl");
+  
+  char *glue = "ue";
+
```
Comment (zvonimir) [inline test/strings/strcat.c]: No need for this new line. You can make these codes tighter in terms of new lines.

File: test/strings/strcat.c [hunk 18]
```diff
@@ -0,0 +1,17 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
```
Comment (zvonimir) [inline test/strings/strcat.c]: strcat and strcat_fail should have the same number or loop unrolls. That holds in general for any _fail variety of a file - their number of unrolls should match.

File: test/strings/strcmp.c [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *smaller = "AAAA";
+  char *bigger = "aaaa";
+
+  int comparison = strcmp(smaller,bigger);
+
```
Comment (zvonimir) [inline test/strings/strcmp.c]: No need for this new line. Like I said, make your files tighter in terms of new lines.

File: test/strings/strtok.c [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+  char *delim = "-";
+
+  char *firstTok = strtok(string,delim);
+  char *nextTok = strtok(NULL,delim);
+
+  assert(strcmp(firstTok,"a") == 0);
+  assert(strcmp(nextTok,"b") == 0);
+
+  return 0;
+
```
Comment (zvonimir) [inline test/strings/strtok.c]: Remove this new line.

File: test/strings/strcat.c [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+
+int main(void) {
+  char brick[10];
+  strcpy(brick,"bl");
+  char *glue = "ue";
+  char *modernart = strcat(brick,glue);
+
```
Comment (zvonimir) [inline test/strings/strcat.c]: Remove this new line.

File: test/strings/strcat_fail.c [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+#include "smack.h"
+
+// @expect error 
+
+int main(void) {
+  char brick[4];
+  strcpy(brick,"bl");
+  char *glue = "u!";
+  char *modernart = strcat(brick,glue);
+
```
Comment (zvonimir) [inline test/strings/strcat_fail.c]: Remove this new line.

File: test/strings/strcat_overflow.c [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+#include "smack.h"
+
+// @flag --memory-safety
+// @expect error
+
+int main(void) {
+  char notLong[3] = "So";
+  char *moreText = "..";
+  char *overflowed = strcat(notLong,moreText);
+
```
Comment (zvonimir) [inline test/strings/strcat_overflow.c]: Remove this new line.

File: test/strings/strchr.c [hunk 24]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *sentence = "a..b";
+  char *withoutFirstWord = strchr(sentence,'.');
+
```
Comment (zvonimir) [inline test/strings/strchr.c]: Remove this new line.

File: test/strings/strchr_fail.c [hunk 25]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *notAWebAddress = "word";
+  char *dotToEnd = strchr(notAWebAddress,'.');
+
```
Comment (zvonimir) [inline test/strings/strchr_fail.c]: Remove this new line.

File: test/strings/strchr_fail.c [hunk 26]
```diff
@@ -0,0 +1,12 @@
...
+  assert(dotToEnd != 0); // or assert(dotToEnd);
```
Comment (zvonimir) [inline test/strings/strchr_fail.c]: Either change the assertion (and remove the comment) or just remove the comment after it.

File: test/strings/strcmp.c [hunk 27]
```diff
@@ -0,0 +1,13 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *smaller = "AAAA";
+  char *bigger = "aaaa";
+  int comparison = strcmp(smaller,bigger);
+
```
Comment (zvonimir) [inline test/strings/strcmp.c]: Remove this new line.

File: test/strings/strcmp_fail.c [hunk 28]
```diff
@@ -0,0 +1,13 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *alpha = "alph";
+  char *zeta = "zeta";
+  int comparison = strcmp(alpha,zeta);
+
```
Comment (zvonimir) [inline test/strings/strcmp_fail.c]: Remove this new line.

File: test/strings/strcpy.c [hunk 29]
```diff
@@ -0,0 +1,15 @@
+#include <strings.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  const char *word = "Roof";
+  char container[5];
+  strcpy(container,word);
+
```
Comment (zvonimir) [inline test/strings/strcpy.c]: Remove this new line.

File: test/strings/strcpy_fail.c [hunk 30]
```diff
@@ -0,0 +1,14 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  const char *word = "made";
+  char container[5];
+  strcpy(container,word);
+
```
Comment (zvonimir) [inline test/strings/strcpy_fail.c]: Remove this new line.

File: test/strings/strcspn.c [hunk 31]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified 
+
+int main(void) {
+  char *string = "ffef";
+  size_t firstE = strcspn(string,"e");
+
```
Comment (zvonimir) [inline test/strings/strcspn.c]: Remove this new line.

File: test/strings/strcspn_fail.c [hunk 32]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *string = "ffff";
+  size_t firstE = strcspn(string,"f");
+
```
Comment (zvonimir) [inline test/strings/strcspn_fail.c]: Remove this new line.

File: test/strings/strlen.c [hunk 33]
```diff
@@ -0,0 +1,12 @@
+#include <stdlib.h>
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char word[] = "Hi!!";
+
```
Comment (zvonimir) [inline test/strings/strlen.c]: Remove this new line.

File: test/strings/strlen_fail.c [hunk 34]
```diff
@@ -0,0 +1,12 @@
+#include <stdlib.h>
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char word[] = "Hi";
+
```
Comment (zvonimir) [inline test/strings/strlen_fail.c]: Remove this new line.

File: test/strings/strncat.c [hunk 35]
```diff
@@ -0,0 +1,14 @@
...
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char dish[40];
+  strcpy(dish,"R");
+  char *milk = "ice";
+  strncat(dish,milk,2);
+
```
Comment (zvonimir) [inline test/strings/strncat.c]: Remove this new line.

File: test/strings/strncat_fail.c [hunk 36]
```diff
@@ -0,0 +1,14 @@
...
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char dish[30];
+  strcpy(dish,"R");
+  char *milk = "ice";
+  strncat(dish,milk,2);
+
```
Comment (zvonimir) [inline test/strings/strncat_fail.c]: Remove this new line.

File: test/strings/strncmp.c [hunk 37]
```diff
@@ -0,0 +1,13 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *str1 = "aaAA";
+  char *str2 = "aaBB";
+
```
Comment (zvonimir) [inline test/strings/strncmp.c]: Remove this new line.

File: test/strings/strncmp_fail.c [hunk 38]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *str1 = "aaAA";
+  char *str2 = "aBBB";
+
```
Comment (zvonimir) [inline test/strings/strncmp_fail.c]: Remove this new line.

File: test/strings/strncmp_toolong.c [hunk 39]
```diff
@@ -0,0 +1,13 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *shorty = "go";
+  char *otherShorty = "go";
+  int comparison = strncmp(shorty, otherShorty, 5);
+
```
Comment (zvonimir) [inline test/strings/strncmp_toolong.c]: Remove this new line.

File: test/strings/strpbrk.c [hunk 40]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *letters = "abcd";
+  char *fromFirstBorC = strpbrk(letters,"cb");
+
```
Comment (zvonimir) [inline test/strings/strpbrk.c]: Remove this new line.

File: test/strings/strpbrk_fail.c [hunk 41]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error 
+
+int main(void) {
+  char *letters = "efgh";
+  char *fromFirstBorC = strpbrk(letters,"bc");
+
```
Comment (zvonimir) [inline test/strings/strpbrk_fail.c]: Remove this new line.

File: test/strings/strrchr.c [hunk 42]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *website = "a..b";
+  char *dotCom = strrchr(website,'.');
+
```
Comment (zvonimir) [inline test/strings/strrchr.c]: Remove this new line.

File: test/strings/strrchr_fail.c [hunk 43]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *website = "a..b";
+  char dotCom = strrchr(website,'.');
+
```
Comment (zvonimir) [inline test/strings/strrchr_fail.c]: Remove this new line.

File: test/strings/strspn.c [hunk 44]
```diff
@@ -0,0 +1,14 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *field = "bbab";
+  size_t firstNonB = strspn(field,"b");
+  size_t end = strspn(field,"ab");
+
```
Comment (zvonimir) [inline test/strings/strspn.c]: Remove this new line.

File: test/strings/strspn_fail.c [hunk 45]
```diff
@@ -0,0 +1,12 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error 
+
+int main(void) {
+  char *field = "bbcb";
+  size_t end = strspn(field,"ab");
+
```
Comment (zvonimir) [inline test/strings/strspn_fail.c]: Remove this new line.

File: test/strings/strstr.c [hunk 46]
```diff
@@ -0,0 +1,13 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *large = "wxyz";
+  char *small = "xy";
+  char *firstOccurrenceOfXY = strstr(large,small);
+
```
Comment (zvonimir) [inline test/strings/strstr.c]: Remove this new line.

File: test/strings/strstr_fail.c [hunk 47]
```diff
@@ -0,0 +1,13 @@
+#include <string.h>
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *large = "lmno";
+  char *small = "xy";
+  char *firstOccurrenceOfXY = strstr(large,small);
+
```
Comment (zvonimir) [inline test/strings/strstr_fail.c]: Remove this new line.

File: test/strings/strtok.c [hunk 48]
```diff
@@ -0,0 +1,16 @@
...
+#include "smack.h"
+
+// @expect verified
+
+int main(void) {
+  char *string = "a-b";
+  char *delim = "-";
+  char *firstTok = strtok(string,delim);
+  char *nextTok = strtok(NULL,delim);
+
```
Comment (zvonimir) [inline test/strings/strtok.c]: Remove this new line.

File: test/strings/strtok_fail.c [hunk 49]
```diff
@@ -0,0 +1,15 @@
...
+#include "smack.h"
+
+// @expect error
+
+int main(void) {
+  char *string = "a--";
+  char *delim = "-";
+  char *firstTok = strtok(string,delim);
+  char *nextTok = strtok(NULL,delim);
+
```
Comment (zvonimir) [inline test/strings/strtok_fail.c]: Remove this new line.

**Pull request merged**